### PR TITLE
feat(MultiValueList): added multiValueListControl and its default APL

### PR DESCRIPTION
**Pull request by Shreyas-vgr**
Adding MultiValueSlot support for skill building using Controls Framework

## Description
- Added MultiValueSlot support by creating a new control `MultiValueListControl` which allows developers to customize and build a skill which allows users to select multiple values from a list of choices.

## Testing
- npm run test success.
- Tested the UX by creating a sample skill described in demo/multiValueListControl to build the InteractionModel and verified the voice inputs and APL touch events.

## Screenshots (if appropriate)
Screen Shot 2021-01-14 at 12 00 12 PM
Screen Shot 2021-01-14 at 12 02 14 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Docs(Add new document content)
- [ ] Translate Docs(Translate document content)

## Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] My commit message follows [Conventional Commit Guideline](https://conventionalcommits.org/)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://https://github.com/alexa/ask-sdk-controls/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts [hunk 1]
```diff
@@ -0,0 +1,1382 @@
...
+    values: MultiValueListStateValue[],
+    input: ControlInput,
+) => true | MultiValueValidationFailure | Promise<true | MultiValueValidationFailure>;
+
+export type AplContent = { document: { [key: string]: any }; dataSource: { [key: string]: any } };
+export type AplContentFunc = (control: MultiValueListControl, input: ControlInput) => AplContent;
+export type AplDocumentPropNewStyle = AplContent | AplContentFunc;
+
+/**
+ * Mapping of action slot values to the behaviors that this control supports.
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControl.ts]: wording: behavior -> capability.

(pls also update the same for other controls as the current wording is confusing)

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts [hunk 2]
```diff
@@ -0,0 +1,1382 @@
...
+ */
+export class MultiValueListControlState implements ControlState {
+    /**
+     * The value.
+     *
+     * If `erMatch = true` the value is a slot value ID for the slot type `this.slotType`.
+     * If `erMatch = false` the value may be an arbitrary string.
+     */
+    value?: MultiValueListStateValue[];
+    /**
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControl.ts]: orphaned comment

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts [hunk 3]
```diff
@@ -0,0 +1,1382 @@
...
+     * Tracks whether the value is an Entity Resolution match.
+     */
+
+    /**
+     * Tracks the most recent elicitation action.
+     *
+     * Note: this isn't cleared immediate after user provides a value as the
+     * value maybe be invalid and has to be re-elicited.  Use
+     * state.activeInitiate to test if the most recent turn was a direct elicitation.
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControl.ts]: state.activeInitiate doesn't appear...   did you mean .lastInitiative?

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts [hunk 4]
```diff
@@ -0,0 +1,1382 @@
...
+ * - Validate the value
+ * - Confirm the value
+ * - Speak the first few options
+ * - Show all the options on APL enabled devices
+ * - Selection of a value using a spoken ordinal, e.g. "The first one".
+ * - Selection of a value using touch screen.
+ *
+ * Intents that can be handled:
+ * - `GeneralControlIntent`: E.g. `"yes, update my name"`
+ * - `{ValueType}_ValueControlIntent`: E.g. "no change it to Elvis".
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControl.ts]: MultiValueControlIntent?

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts [hunk 5]
```diff
@@ -0,0 +1,1382 @@
...
+            return falseIfGuardFailed(e);
+        }
+    }
+
+    private async handleAddWithValue(
+        input: ControlInput,
+        resultBuilder: ControlResultBuilder,
+    ): Promise<void> {
+        const slotValues = InputUtil.getMultiValueResolution(input);
+        let values = this.generateSlotValues(slotValues);
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControl.ts]: generateSlotValue is an odd name.  getSlotValues?

File: src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                                paddingLeft: '@spacingMedium',
+                                direction: 'row',
+                                items: [
+                                    {
+                                        type: 'AlexaTextList',
+                                        width: '50vw',
+                                        height: '80vh',
+                                        touchForward: true,
+                                        headerBackButton: false,
+                                        primaryAction: {
```
Comment (mikelid) [inline src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts]: suggest trialing "disable" functionality so that user can't press more buttons before the round trip completes and cause race conditions.
Although, if the operations are all safe to race then it might be a feature.

I'm concerned about remove which is currently implemented via Ordinals which are race-sensitive.  If removal changed to being by (value, ordinal), then races might be OK.  The server logic would be to remove the first matching item at ordinal or lower to allow for races such as "delete item #1" followed immediately by "delete item #2" which would need to be processed as a second instance of delete item#1.

tl;dr:   avoiding races is safest but in this particular case you might be able to make races safe and could be a nice feature?

**Review by mikelid — COMMENTED** [review]
Awesome!  Very clean code and a great new Control in our library.

File: src/commonControls/LanguageStrings.ts [hunk 7]
```diff
@@ -107,6 +107,9 @@ export const defaultI18nResources: Resource = {
             // ListControl Runtime
             LIST_CONTROL_DEFAULT_PROMPT_VALUE_SET: 'OK, {{value}}.',
             LIST_CONTROL_DEFAULT_PROMPT_VALUE_CHANGED: 'OK, I changed it to {{value}}.',
+            LIST_CONTROL_DEFAULT_PROMPT_VALUE_ADD: 'OK, added {{value}}.',
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/LanguageStrings.ts:110]: Yes, these are used by ListControl to add/remove values.
I think it's better to keep them separate to change default prompts on controls individually if required?

File: src/commonControls/LanguageStrings.ts [hunk 8]
```diff
@@ -935,6 +991,41 @@ export const defaultI18nResources: Resource = {
...
+                        name: {
+                            value: 'add',
+                            synonyms: ['add'],
+                        },
+                    },
+                    {
+                        id: $.Action.Remove,
+                        name: {
+                            value: 'remove',
+                            synonyms: ['remove', 'delete'],
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/LanguageStrings.ts:1005]: Agreed, consolidated and deleted the `$.Action.Delete`. Currently on existing controls both have same actions.

File: src/commonControls/LanguageStrings.ts [hunk 9]
```diff
@@ -167,6 +178,51 @@ export const defaultI18nResources: Resource = {
...
+            MULTI_VALUE_LIST_CONTROL_DEFAULT_REPROMPT_REQUEST_VALUE:
+                'What is your selection? Some suggestions are {{suggestions}}.',
+            MULTI_VALUE_LIST_CONTROL_DEFAULT_REPROMPT_REQUEST_REMOVED_VALUE:
+                'What value do you want to remove? Some suggestions are {{suggestions}}.',
+            MULTI_VALUE_LIST_CONTROL_DEFAULT_REPROMPT_GENERAL_REQUEST_REMOVED_VALUE:
+                'What value do you want to remove?',
+            MULTI_VALUE_LIST_CONTROL_DEFAULT_REPROMPT_CONFIRM_VALUE: 'Ok I have {{value}}. Is that all?',
+            MULTI_VALUE_LIST_CONTROL_DEFAULT_REPROMPT_VALUE_AFFIRMED: 'Great.',
+            MULTIVALUELIST_CONTROL_DEFAULT_APL_HEADER_TITLE: 'Create your List',
+            MULTIVALUELIST_CONTROL_DEFAULT_APL_HEADER_SUBTITLE: 'Say an item or touch it to add it your list',
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/LanguageStrings.ts:224]: Yup, will update screenshot after all fixes and update with latest changes.

File: src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+                                                    ],
+                                                },
+                                            ],
+                                        },
+                                        listItems: '${payload.choices.listItems}',
+                                    },
+                                    {
+                                        type: 'AlexaTextList',
+                                        touchForward: true,
+                                        headerTitle: 'Your selections',
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts]: Updated

File: src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+                            },
+                            {
+                                type: 'AlexaButton',
+                                buttonText: 'Done',
+                                id: 'actionComplete',
+                                primaryAction: {
+                                    type: 'SendEvent',
+                                    arguments: ['${payload.general.controlId}', 'Complete'],
+                                },
+                                right: '10vw',
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts]: Updated the APL document in latest commit

File: src/intents/MultiValueControlIntent.ts [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+ */
+export interface MultiValueControlIntentSlots {
+    feedback?: string;
+    action?: string;
+    target?: string;
+    [key: string]: string[] | string | undefined;
+}
+
+export interface MultiValueSlot {
+    [key: string]: string | boolean;
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/intents/MultiValueControlIntent.ts:31]: To track erMatch status
`MultiValueSlot  = { slotValue: string; isEntityResolutionMatch: boolean; }`

File: src/systemActs/PayloadTypes.ts [hunk 13]
```diff
@@ -113,6 +165,18 @@ export interface RequestChangedValueByListPayload {
     renderedChoices?: string;
 }
 
+/**
+ * Payload for RequestRemovedValueByListAct
+ */
+export interface RequestRemovedValueByListActPayload {
+    availableChoicesFromActivePage: string[];
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/systemActs/PayloadTypes.ts:172]: So in the ListControl/MultiValueListControl the list of choices can be a lot(exceeding 5+) to be spoken by Alexa on request of an action with suggestions, `availableChoicesFromActivePage` tracks the reduced list spoken from prompts as suggestions.

This is customized by the props `pageSize` which defaults to 3 choices. Hence I store for both the limited choices and entire list on payloads.

File: src/commonControls/MultiValueListControl/MultiValueListControl.ts (hunk 5 above)
Comment (Shreyas-vgr) [reply src/commonControls/MultiValueListControl/MultiValueListControl.ts]: Updated to `getSlotValues()`

File: src/intents/MultiValueControlIntent.ts [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        this.filteredValueSlotType = filteredValueSlotType ?? valueSlotType;
+        this.name = MultiValueControlIntent.intentName(valueSlotType);
+    }
+
+    /**
+     * Generates the intent name of a specialized `MultiValueControlIntent`.
+     *
+     * Example:
+     * - The intent name for a `MultiValueControlIntent` that conveys an
+     *   `AMAZON.NUMBER` is `AMAZON_NUMBER_MultiValueControlIntent`.
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/intents/MultiValueControlIntent.ts:172]: Nope, I tested with a sample skill and I didn't see any conflicts.

File: src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts (hunk 6 above)
Comment (Shreyas-vgr) [reply src/commonControls/MultiValueListControl/MultiValueListControlAPL.ts]: As discussed, will set `disabled: boolean` property on list of choices when selected and also will disable swipe action on items when a single item is removed to avoid race conditions.

File: demo/MultiValueListControl/src/buildInteractionModel.ts [hunk 15]
```diff
@@ -31,41 +34,41 @@ export namespace MultiValueListDemoIM {
...
                     name: {
-                        value: 'iWatch',
-                        synonyms: ['watch', 'wrist wear'],
+                        value: 'Eggs',
+                        synonyms: ['egg', 'eggs'],
                     },
                 },
                 {
-                    id: 'iPhone',
+                    id: 'Ceral',
```
Comment (mikelid) [inline demo/MultiValueListControl/src/buildInteractionModel.ts]: nit: Ceral -> Cereal

File: demo/MultiValueListControl/src/buildInteractionModel.ts [hunk 16]
```diff
@@ -51,7 +51,7 @@ export namespace MultiValueListDemoIM {
                     },
                 },
                 {
-                    id: 'Ceral',
+                    id: 'Cereal',
```
Comment (Shreyas-vgr) [inline demo/MultiValueListControl/src/buildInteractionModel.ts:54]: @mikelid, It's fixed already :)

File: src/commonControls/LanguageStrings.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikelid) [reply src/commonControls/LanguageStrings.ts:110]: I don't think they are... list control has set/change capabilities but not add/remove.  I don't see them referenced by ListControl

File: demo/MultiValueListControl/src/buildInteractionModel.ts (hunk 15 above)
Comment (Shreyas-vgr) [reply demo/MultiValueListControl/src/buildInteractionModel.ts]: I've had fixed this is recent commit, maybe you didn't look at latest file

**Review by mikelid — CHANGES_REQUESTED** [review]
Just one open issue about string config then good to go!

**Review by mikelid — APPROVED** [review]

File: src/commonControls/LanguageStrings.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/commonControls/LanguageStrings.ts]: Ah yes, sorry for the confusion, removed these constants in latest commit. 👍
